File: archive/python/src/loaders/text_loader.py
import re
from src.loaders.base import DocumentLoader
from src.models import DocumentPage, TextLine
MARKDOWN_HEADING = re.compile(r"^(#{1,6})\s+(.*\S)\s*$")
class TextDocumentLoader(DocumentLoader):
    extensions = (".txt", ".md", ".markdown")
    def load(self) -> list[DocumentPage]:
        self._validate()
        raw = self.file_path.read_text(encoding = "utf-8", errors = "replace")
        lines = []
        for line in raw.splitlines():
            stripped = line.strip()
            if not stripped:
                continue
            heading = MARKDOWN_HEADING.match(stripped)
            if heading:
                lines.append(
                    TextLine(
                        text = heading.group(2),
                        page_number = 1,
                        heading_level = len(heading.group(1)),
                    )
                )
            else:
                lines.append(TextLine(text = stripped, page_number = 1))
        return [
            DocumentPage(
                text = "\n".join(line.text for line in lines),
                page_number = 1,
                source = self.file_path.name,
                lines = lines,
            )
        ]

File: src/loaders/textLoader.ts
import { readFile } from "node:fs/promises";
import { basename } from "node:path";
import { DocumentLoader } from "./base.ts";
import type { DocumentPage, TextLine } from "../models.ts";

const MARKDOWN_HEADING = /^(#{1,6})\s+(.*\S)\s*$/;

export class TextDocumentLoader extends DocumentLoader {
  readonly extensions = [".txt", ".md", ".markdown"] as const;

  async load(filePath: string): Promise<DocumentPage[]> {
    await this.validate(filePath);
    const raw = await readFile(filePath, "utf8");

    const lines: TextLine[] = [];
    // A blank line is the paragraph break in a text file - the one structural
    // signal the format has. Record it rather than dropping it.
    let blankBefore = false;
    for (const line of raw.split(/\r?\n/)) {
      const stripped = line.trim();
      if (!stripped) {
        blankBefore = true;
        continue;
      }
      const heading = MARKDOWN_HEADING.exec(stripped);
      lines.push(
        heading
          ? {
              text: heading[2]!,
              pageNumber: 1,
              headingLevel: heading[1]!.length,
              breakBefore: true,
            }
          : { text: stripped, pageNumber: 1, breakBefore: blankBefore },
      );
      blankBefore = false;
    }

    // A text file has no pages, so it is one page.
    return [
      {
        text: lines.map((line) => line.text).join("\n"),
        pageNumber: 1,
        source: basename(filePath),
        lines,
      },
    ];
  }
}
